accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,484 @@
+    /// to start recording measurements.
+    /// </summary>
+
+    public abstract class Instrument
+    {
+        /// <summary>
+        /// Protected constructor to initialize the common instrument properties.
+        /// </summary>
+        protected Instrument(Meter meter, string name, string? description, string? unit) { throw null; }
+
+        /// <summary>
+        /// Publish is to allow activating the instrument to start recording measurements and to allow
+        /// listeners to start listening to such measurements.
+        /// </summary>
+        protected void Publish() { throw null; }
+
+        /// <summary>
+        /// Getters to retrieve the properties that the instrument is created with.
+        /// </summary>
+        public Meter Meter { get; }
+        public string Name { get; }
+        public string? Description { get; }
+        public string? Unit { get; }
+
+        /// <summary>
+        /// A property tells if a listener is listening to this instrument measurement recording.
+        /// </summary>
+        public bool Enabled => throw null;
+
+        /// <summary>
+        /// A property tells if the instrument is a regular instrument or an observable instrument.
+        /// </summary>
+        public virtual bool IsObservable => throw null;
+    }
+}
+```
+
+### Non-observable Instrument Base Class
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Instrument<T> is the base class from which all non-observable instruments will inherit from.
+    /// Mainly It'll support the CLS compliant numerical types
+    /// </summary>
+    public abstract class Instrument<T> : Instrument where T : unmanaged
+    {
+        /// <summary>
+        /// Protected constructor to create the instrument with the common properties.
+        /// </summary>
+        protected Instrument(Meter meter, string name, string? description, string? unit) :
+                        base(meter, name, description, unit) { throw null; }
+
+        /// <summary>
+        /// Record measurement overloads allowing passing different numbers of tags.
+        /// </summary>
+
+        protected void RecordMeasurement(T val) { throw null; }
+
+        protected void RecordMeasurement(
+                            T measurement,
+                            KeyValuePair<string, object?> tag1) { throw null; }
+
+        protected void RecordMeasurement(
+                            T measurement,
+                            KeyValuePair<string, object?> tag1,
+                            KeyValuePair<string, object?> tag2) { throw null; }
+
+        protected void RecordMeasurement(
+                            T measurement,
+                            KeyValuePair<string, object?> tag1,
+                            KeyValuePair<string, object?> tag2,
+                            KeyValuePair<string, object?> tag3) { throw null; }
```

Comment:
Given our Listener doesn't also support these 1, 2, 3 KVP prototypes. Do we really need these 1, 2, 3 KVP prototypes?
Also, here is the protected versions, so I assume derived classes can convert to ReadOnlySpan<> for us.

Reply:
The 1, 2, 3 KVPs overloads are needed to avoid creating array parameter.